plugins/out_cloudwatch_logs/cloudwatch_api.c

```diff
@@ -786,6 +1019,175 @@ int pack_emf_payload(struct flb_cloudwatch *ctx,
     return 0;
 }
 
+static char* find_fallback_environment(struct flb_cloudwatch *ctx, entity *entity) {
+    if(!ctx->add_entity || entity == NULL) {
+        return NULL;
+    }
+    char *fallback_env = NULL;
+    int ret;
+    /*
+     * Possible fallback environments:
+     * 1. eks:cluster-name/namespace
+     * 2. k8s:cluster-name/namespace
+     */
+    if (entity->attributes->platform_type != NULL && entity->attributes->cluster_name != NULL && entity->attributes->namespace != NULL) {
+        /* Calculate required length */
+        /* Add 3 for ':' '/' and null terminator */
+        size_t len = strlen(entity->attributes->platform_type) +
+                    strlen(entity->attributes->cluster_name) +
+                    strlen(entity->attributes->namespace) + 3;
+
+        fallback_env = flb_malloc(len);
+        if (!fallback_env) {
+            return NULL;
+        }
+
+        /* Use snprintf for cross-platform compatibility */
+        ret = snprintf(fallback_env, len, "%s:%s/%s", entity->attributes->platform_type, entity->attributes->cluster_name, entity->attributes->namespace);
+        if (ret < 0 || ret >= len) {
+            flb_free(fallback_env);
+            return NULL;
+        }
+
+        return fallback_env;
+    }
+    return NULL;
+}
+
+void parse_entity(struct flb_cloudwatch *ctx, entity *entity, msgpack_object map, int map_size) {
+    int i,j;
+    msgpack_object key, kube_key;
+    msgpack_object val, kube_val;
+
+    int val_map_size;
+    for(i=0; i < map_size; i++) {
```

Comment:
this is a good candidate to be refactored by using the record accessor interface:

- end-user ref: https://docs.fluentbit.io/manual/administration/configuring-fluent-bit/classic-mode/record-accessor?q=record
- context creation example: https://github.com/fluent/fluent-bit/blob/master/plugins/out_loki/loki.c#L1005
- value lookup example: https://github.com/fluent/fluent-bit/blob/309a3ee2e18a0ab8c7ef83a98a6b4b8baa4356f5/plugins/out_loki/loki.c#L595

in general for any other key lookup it's better to simplify the logic